### PR TITLE
Guide to using JSDoc

### PR DESCRIPTION
**Pull request by nmashton**

### CODE REVIEW

File: source/frontend-docs.rst [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    npm install jsdoc --save-dev
+
+For most projects, you will want to create a configuration file that will
+tell JSDoc such things as which files to ignore, which plugins to use, etc.
+Full instructions can be found `on the JSDoc site <http://usejsdoc.org/about-configuring-jsdoc.html>`_.
+
+Once the configuration file is set up, you can generate a JSDoc site from your
+JS by running this command::
+
+    node ./node_modules/jsdoc/jsdoc.js -c name-of-config-file.json -r path/to/js -d output/dir
```
Comment (ironfroggy) [inline source/frontend-docs.rst]: where should i output it?
where is this configuration file and what should i name it?

File: source/frontend-docs.rst [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+If your project uses ECMAScript 2015 with Babel, you need to use
+the ``jsdoc-babel`` plugin::
+
+    npm install jsdoc@3.3.x jsdoc-babel --save-dev
+
+You will then need to set up your configuration file to include ``jsdoc-babel``::
```
Comment (ironfroggy) [inline source/frontend-docs.rst]: Which configuration file? JSDoc's or Babel's?

File: source/frontend-docs.rst (hunk 1 above)
Comment (ironfroggy) [inline source/frontend-docs.rst]: shouldn't this be the jsdoc script at `node_modules/.bin/jsdoc` or just `jsdoc` if your node scripts are in `PATH`?

File: source/frontend-docs.rst [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+JSDoc is not a library and doesn't expose a full-featured API. For that
+reason, to use JSDoc from within your build process, it should be treated as
+just another command line program to be run.
+
+In Gulp, for instance, you can run the above command by means of Node's
+``child_process``, like so::
+
+    gulp.task('jsdoc', function (cb) {
+      child_exec('node ./node_modules/jsdoc/jsdoc.js -c ./name-of-config-file.json -r ./path/to/js -d ./output/dir', undefined, cb);
+    });
```
Comment (ironfroggy) [inline source/frontend-docs.rst]: Our documentation outlines getting all projects using gulp. so maybe assume that and don't show the raw command above, just outline how to add it to the gulpfile we expect projects have.

Comment (ironfroggy) [inline source/frontend-docs.rst]: And include the necessary `require()`

File: source/frontend-docs.rst [hunk 4]
```diff
@@ -0,0 +1,218 @@
+Front-end documentation with JSDoc
+##################################
+
+Writing maintainable and understandable front-end JavaScript code calls for writing
+particularly thorough documentation. JavaScript is an unusually flexible language in which
+protocols and expectations are rarely if ever obvious without carefully reading
+the code. It's also a language with a huge and diverse set of libraries where
+standards and conventions are in a constant state of flux. The only way to
+ensure that the developer who inherits your front-end code tomorrow or a year
+from now will understand what your program does is to carefully document your code,
```
Comment (dpoirier) [inline source/frontend-docs.rst]: (who might be you!)

File: source/frontend-docs.rst [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+used in major JS projects. Whether or not you use JSDoc to generate an API
+docs site for your code, following the JSDoc conventions is a good way to ensure
+that your comments are as informative and immediately comprehensible as possible.
+
+JSDoc installation and basic setup
+==================================
+
+JSDoc can be added to an existing project as an npm dependency. To take advantage
+of JSDoc 3.4's native support for ECMAScript 2015, install a version of JSDoc
+later than 3.4.0::
```
Comment (dpoirier) [inline source/frontend-docs.rst]: later than, or at least?

File: source/frontend-docs.rst [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+    npm install jsdoc^3.4.0 --save-dev
+
+For most projects, you will want to create a configuration file that will
+tell JSDoc such things as which files to ignore, which plugins to use, etc.
+Full instructions can be found `on the JSDoc site <http://usejsdoc.org/about-configuring-jsdoc.html>`_.
+
+One important step is to `create a configuration file <http://usejsdoc.org/about-configuring-jsdoc.html#configuration-file>`_.
```
Comment (dpoirier) [inline source/frontend-docs.rst]: This sentence seems to repeat the previous paragraph?

File: source/frontend-docs.rst [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+project directory (alongside other config files like ``.gitignore``, ``.babelrc``, etc).
+This configuration file specifies the plugins JSDoc will use, among other things.
+
+In your config file, you can set up the source and destination directories
+that JSDoc will look through as it crawls your JS code and builds the documentation
+site. In projects created with our project template, the source dir will look
+like ``project_name/static/js``; we recommend ``docs/js`` as the output dir.
+You'll also want to enable the ``recurse`` option so that your entire source tree
+will be scanned. If your JavaScript build output is located in the same directory
+as your source, you will also want to throw in an ``excludePattern``. Including
```
Comment (dpoirier) [inline source/frontend-docs.rst]: but that would go against our recommendations?

Comment (dpoirier) [inline source/frontend-docs.rst]: oh - this is referring to babelifying the javascript, not creating docs from it.  I was thinking about building documentation when I read "JavaScript build output" - maybe we could be a bit more explicit here about which build we're referring to?

File: source/frontend-docs.rst [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+==============================
+
+JSDoc offers a very rich selection of tags and other features and a great deal
+of flexibility in how it can be used. We recommend adhering to at least these
+conventions:
+
+* Create a doc comment for every module, class, function / method, and event.
+* For every function / method doc comment, include:
+    * A prose description of the function's purpose
+    * ``@param`` tags each expected parameter, including:
```
Comment (dpoirier) [inline source/frontend-docs.rst]: s/tags each/tags for each/ ?

File: source/frontend-docs.rst [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        * name
+        * type
+        * prose description
+    * For Object-type parameters, include separate ``@param`` entries for each
+      property expected on the Object
+    * ``@returns`` tag giving the function's return value, including:
+        * type
+        * prose description
```
Comment (dpoirier) [inline source/frontend-docs.rst]: Should we include doc on the properties that will be included on a returned Object, like we do for incoming parameters?

File: source/frontend-docs.rst [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+constructor. This comment includes a brief prose description of
+the class's nature; it explicitly identifies the function as a
+constructor using ``@constructs``; it spells out the type of its single
+required parameter, ``options``, and the attribute ``id`` that it
+requires; and it uses ``@see`` to link to the docs for the method
+``appSetup`` which does the constructor's heavy lifting::
+
+    export default class ShowMatches extends Model {
+      /**
+       * Create a new app state model.
```
Comment (dpoirier) [inline source/frontend-docs.rst]: Is this the documentation for the class or the constructor?  The prose description reads like it's just documenting the constructor, but if JSDoc is going to use this comment for the whole module, maybe it should be broadened.

File: source/frontend-docs.rst [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+     * @param {module:models/PossibleMatch.PossibleMatch[]} matches - set of
+     *     matches to compare
+     * @fires ShowMatches#event:change:comparing_matches
+     * @see module:views/PossibleMatches
+     */
+    compareMatches (models) {
+      this.set('comparing_matches', new PossibleMatchesCollection(models));
+    }
+
+**Important**: note that ``@method`` gives the name of the class and the
```
Comment (dpoirier) [inline source/frontend-docs.rst]: Sphinx has nice built-in support for formatting warning boxes so you don't have to format them yourself - see "Admonitions" here: http://www.sphinx-doc.org/en/stable/rest.html#directives

File: source/frontend-docs.rst [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+    * ``@fires`` tags indicating events fired within the function body
+* If a function is a callback:
+    * Declare its doc comment with ``@callback``
+    * Specify the event(s) it's attached to with ``@listens``
+* Include ``@see`` references wherever another part of the code provides useful
+  context for interpreting the comment.
+
+For the full details of what JSDoc is capable of, see `the JSDoc website <http://usejsdoc.org/>`_.
+
+Examples
```
Comment (ironfroggy) [inline source/frontend-docs.rst]: I'd rename this like Additional Usage or something. "Examples" might often get skipped over.

**Comment by ironfroggy** [comment]
+1

**Pull request merged**